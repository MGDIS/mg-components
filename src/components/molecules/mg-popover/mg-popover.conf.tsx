/**
 * List of all possibles placements
 */
export const placements = [
  'auto',
  'auto-start',
  'auto-end',
  'top',
  'top-start',
  'top-end',
  'bottom',
  'bottom-start',
  'bottom-end',
  'right',
  'right-start',
  'right-end',
  'left',
  'left-start',
  'left-end',
] as const;

/**
 * Placement type from placements
 */
export type Placement = typeof placements[number];
