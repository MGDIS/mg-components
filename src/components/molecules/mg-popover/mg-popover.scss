$position-size: 6px;

@mixin arrow-placement($position, $shadow-x, $shadow-y) {
  #{$position}: calc(-1 * #{$position-size});
  filter: drop-shadow(#{$shadow-x} #{$shadow-y} 5px hsla(var(--color-dark), 0.15));
}

:host {
  --mg-card-padding: 1.5rem;
  --mg-card-border: none;
  --mg-card-border-radius: 0.5rem;
  --mg-card-background: hsl(var(--mg-popover-background-color));
  position: relative;

  display: contents;
  &::slotted(*) {
    display: inline-block;
  }
}

.mg-popover {
  display: none;
  z-index: 2000;
  color: hsl(var(--mg-popover-font-color));

  mg-card {
    z-index: -1;
  }

  &[data-show] {
    display: block;
  }

  mg-button {
    float: right;
    margin-right: calc(0rem - (var(--default-size) - var(--mg-icon-regular-size)) / 2); // Without unit (rem) the value can not be negative
    margin-top: calc(0rem - (var(--default-size) - var(--mg-popover-title-font-size) * var(--line-height)) / 2);
    margin-left: 3rem;
  }

  &__arrow {
    visibility: hidden;
    &,
    &::before {
      // Display
      position: absolute;
      width: 2rem;
      height: 2rem;
      z-index: 2000;
      // Decoration
      background: hsl(var(--mg-popover-background-color));
    }
    &::before {
      visibility: visible;
      content: '';
      transform: rotate(45deg);
    }
  }

  &[data-popper-placement^='top'] > &__arrow {
    @include arrow-placement('bottom', 0, calc(#{$position-size} * 2));
  }

  &[data-popper-placement^='bottom'] > &__arrow {
    @include arrow-placement('top', 0, calc((-1 * #{$position-size}) * 2));
  }

  &[data-popper-placement^='left'] > &__arrow {
    @include arrow-placement('right', calc(#{$position-size} * 2), 0);
  }

  &[data-popper-placement^='right'] > &__arrow {
    @include arrow-placement('left', calc((-1 * #{$position-size}) * 2), 0);
  }

  &__title ::slotted([slot='title']) {
    margin: 0 0 1rem !important;
    font-size: var(--mg-popover-title-font-size) !important;
    font-weight: 600;
  }
}
