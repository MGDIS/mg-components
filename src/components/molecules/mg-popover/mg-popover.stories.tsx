import { h } from '@stencil/core';

export default {
  component: 'mg-popover',
  title: 'Atoms/mg-popover',
  parameters: {
    layout: 'centered',
    docs: { iframeHeight: 600 },
  },
};

/**
 * 1. camelCase arguments must be written in the template, for exemple labelOnTop must be placed in the template as label-on-top={args.labelOnTop}
 * 2. boolean arguments with a default true value must be added like display-character-left={args.displayCharacterLeft ? 'true' : 'false'}
 */

const Template = args => {
  // Extract slot so it won't be render as an attribute
  const slotTitle = args.slotTitle;
  delete args.slotTitle;
  const slotContent = args.slotContent;
  delete args.slotContent;
  const closeButton = args.closeButton;
  delete args.closeButton;
  // return element
  return (
    <mg-popover {...args} close-button={closeButton}>
      <mg-button>Button</mg-button>
      {slotTitle && <h2 slot="title">{slotTitle}</h2>}
      {slotContent && <p slot="content">{slotContent}</p>}
    </mg-popover>
  );
};

export const MgPopover = Template.bind({});

MgPopover.args = {
  slotTitle: `Blu bli blo bla`,
  slotContent: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.`,
  identifier: 'identifier',
  closeButton: false,
  disabled: false,
  display: false,
};
