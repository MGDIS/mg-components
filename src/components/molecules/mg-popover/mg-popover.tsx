import { Component, Element, Host, h, Prop, Watch, EventEmitter, Event } from '@stencil/core';
import { createID, isTagName } from '../../../utils/components.utils';
import { Instance as PopperInstance, createPopper, Placement } from '@popperjs/core';
import { initLocales } from '../../../locales';

@Component({
  tag: 'mg-popover',
  styleUrl: 'mg-popover.scss',
  shadow: true,
})
export class MgPopover {
  /************
   * Internal *
   ************/

  private popper: PopperInstance;
  private popover: HTMLElement;
  private closeButtonId = '';

  // Locales
  private messages;

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgPopoverElement;

  /**
   * Sets an `id` attribute.
   * Needed by the input for accessibility `aria-decribedby`.
   */
  @Prop() identifier: string = createID('mg-popover');

  /**
   * Popover placement
   */
  @Prop() placement: Placement = 'bottom';

  /**
   * Define if popover has a cross button
   */
  @Prop() closeButton = false;

  /**
   * Display popover
   */
  @Prop({ mutable: true }) display = false;
  @Watch('display')
  handleDisplay(newValue: boolean): void {
    if (newValue) {
      this.show();
    } else {
      this.hide();
    }
    this.displayChange.emit(newValue);
  }

  /**
   * Disable popover
   */
  @Prop({ mutable: true }) disabled = false;

  /**
   * Emited event when display value change
   */
  @Event({ eventName: 'display-change' }) displayChange: EventEmitter<boolean>;

  /**
   * Check if clicked outside of component
   *
   * @param {MouseEvent} event mouse event
   * @returns {void}
   */
  private clickOutside = (event: MouseEvent & { target: HTMLElement }): void => {
    if (!this.disabled && event.target.closest(`mg-popover`) !== this.element) {
      this.display = false;
    }
  };

  /**
   * Show popover
   *
   * @returns {void}
   */
  private show = (): void => {
    // Make the popover visible
    this.popover.setAttribute('data-show', '');
    // Enable the event listeners
    this.popper.setOptions(options => ({
      ...options,
      modifiers: [...options.modifiers, { name: 'eventListeners', enabled: true }],
    }));
    // Update its position
    this.popper.update();
    // hide when click outside
    // setTimeout is used to prevent event to trigger after creation
    setTimeout(() => {
      document.addEventListener('click', this.clickOutside, false);
    }, 0);
  };

  /**
   * Hide popover
   *
   * @returns {void}
   */
  private hide = (): void => {
    // Hide the popover
    this.popover.removeAttribute('data-show');
    // Disable the event listeners
    this.popper.setOptions(options => ({
      ...options,
      modifiers: [...options.modifiers, { name: 'eventListeners', enabled: false }],
    }));
    // Remove event listener
    document.removeEventListener('click', this.clickOutside, false);
  };

  /**
   * Handle action for close button
   *
   * @returns {void}
   */
  private handleCloseButton = (): void => {
    this.display = false;
  };

  /*************
   * Lifecycle *
   *************/

  /**
   * Check if component props are well configured on init
   *
   * @returns {void} timeout
   */
  componentWillLoad(): void {
    // Get locales
    this.messages = initLocales(this.element).messages;
  }

  /**
   * Check if component props are well configured on init
   *
   * @returns {void}
   */
  componentDidLoad(): void {
    const headingTags = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'];

    const slottedTitleElement = this.element.querySelector('[slot="title"]');
    if (slottedTitleElement && !isTagName(slottedTitleElement, headingTags)) {
      throw new Error(`<mg-popover> Slotted title must be a heading: ${headingTags.join(', ')}`);
    }

    // Set close button id
    if (this.closeButton) {
      this.closeButtonId = `${this.identifier}-close-button`;
    }

    // Get popover content
    this.popover = this.element.shadowRoot.querySelector(`#${this.identifier}`);

    //Get interactive element (first element without slot attribute)
    const interactiveElement = this.element.querySelector(':not([slot])') as HTMLElement;
    // Add aria attributes
    interactiveElement.setAttribute('aria-controls', this.identifier);
    interactiveElement.setAttribute('aria-expanded', `${this.display}`);

    // Create popperjs popover
    this.popper = createPopper(interactiveElement, this.popover, {
      placement: this.placement,
      strategy: 'fixed',
      modifiers: [
        {
          name: 'offset',
          options: {
            offset: [0, 10],
          },
        },
      ],
    });

    // Add events to toggle display
    interactiveElement.addEventListener('click', () => {
      if (!this.disabled) this.display = !this.display;
    });

    // Add events to hide popover
    this.element.addEventListener('keydown', e => {
      if (!this.disabled && e.code === 'Escape') this.display = false;
    });

    this.handleDisplay(this.display);
  }

  /**
   * Render
   *
   * @returns {HTMLElement} HTML Element
   */
  render(): HTMLElement {
    return (
      <Host>
        <slot></slot>
        <div id={this.identifier} class="mg-popover">
          {this.closeButton && (
            <mg-button identifier={this.closeButtonId} is-icon variant="flat" label={this.messages.general.close} onClick={this.handleCloseButton}>
              <mg-icon icon="cross"></mg-icon>
            </mg-button>
          )}
          <div class="mg-popover__title">
            <slot name="title"></slot>
          </div>
          <slot name="content"></slot>
          <div class="mg-popover__arrow" data-popper-arrow></div>
        </div>
      </Host>
    );
  }
}
