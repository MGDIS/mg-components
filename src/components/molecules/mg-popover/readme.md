## Design 

Un popover est plus enrichi qu'une tooltip mais moins bloquant qu'une modale

**RUX-01** : la fenêtre est affichée au clic sur son déclencheur

**RUX-02** : la fenêtre est affichée à côté de son déclencheur, le plus souvent en dessous. Un triangle (en css) fait le lien entre le déclencheur et la fenêtre.

**RUX-03** : la fenêtre se ferme

au clic sur son déclencheur ou
au clic en dehors de la fenêtre ou
à la pression de la touche ECHAP
au clic sur le bouton de fermeture en haut à droite

**RUX-04** : la taille de la fenêtre est déterminée par le contenu (prévoir une taille max pour éviter des écueils)

**RUX-05** : il est possible de masquer le header (titre + bouton fermer)

![](/img/components/mg-popover/2affichages.png)

### Formes

![](/img/components/mg-popover/formes-popover.png)

### Fonts

![](/img/components/mg-popover/font-popover.png)

### Espacements

![](/img/components/mg-popover/espacements-popover-1.png)

#### Entre les bords et le contenu

![](/img/components/mg-popover/espacements-popover-2.png)

#### Entre le header et le contenu

![](/img/components/mg-popover/espacements-popover-3.png)

#### Entre le header et le bouton fermé

![](/img/components/mg-popover/espacements-popover-4.png)

### Alignements

![](/img/components/mg-popover/alignements-popover.png)

### Positionnements

![](/img/components/mg-popover/positionement-popover.png)
![](/img/components/mg-popover/positionement-triangle.png)

### Couleurs

![](/img/components/mg-popover/couleurs-popover.png)

### Tailles

![](/img/components/mg-popover/Tailles-popover.png)

### Shadow

![](/img/components/mg-popover/shadow-popover.png)

<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description                                                                      | Type                                                                                                                                                                                                         | Default                  |
| ------------- | -------------- | -------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------ |
| `closeButton` | `close-button` | Define if popover has a cross button                                             | `boolean`                                                                                                                                                                                                    | `false`                  |
| `disabled`    | `disabled`     | Disable popover                                                                  | `boolean`                                                                                                                                                                                                    | `false`                  |
| `display`     | `display`      | Display popover                                                                  | `boolean`                                                                                                                                                                                                    | `false`                  |
| `identifier`  | `identifier`   | Sets an `id` attribute. Needed by the input for accessibility `aria-decribedby`. | `string`                                                                                                                                                                                                     | `createID('mg-popover')` |
| `placement`   | `placement`    | Popover placement                                                                | `"auto" \| "auto-end" \| "auto-start" \| "bottom" \| "bottom-end" \| "bottom-start" \| "left" \| "left-end" \| "left-start" \| "right" \| "right-end" \| "right-start" \| "top" \| "top-end" \| "top-start"` | `'bottom'`               |


## Dependencies

### Depends on

- [mg-button](../../atoms/mg-button)
- [mg-icon](../../atoms/mg-icon)

### Graph
```mermaid
graph TD;
  mg-popover --> mg-button
  mg-popover --> mg-icon
  mg-button --> mg-icon
  style mg-popover fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
