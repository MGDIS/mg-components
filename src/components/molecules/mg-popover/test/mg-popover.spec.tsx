import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgPopover } from '../mg-popover';
import { MgButton } from '../../../atoms/mg-button/mg-button';
import { placements } from '../mg-popover.conf';

// fix popper console.error in test
// it is generated in @popperjs/core/dist/cjs/popper.js l.1859
// this is due to internal function isHTMLElement(), so we can not mock it directly.
// this function check if test DOM element mockHTMLElement instance is 'instanceof HTMLElement'
// so we only override the console.error side effect for this error
const errorFunction = console.error;
const mock = jest.spyOn(console, 'error');
mock.mockImplementation(error => {
  const compareWith = 'Popper: "arrow" element must be an HTMLElement (not an SVGElement). To use an SVG arrow, wrap it in an HTMLElement that will be used as the arrow.';
  if (error !== compareWith) {
    errorFunction(error);
  }
});

const getPage = (args, element) =>
  newSpecPage({
    components: [MgPopover, MgButton],
    template: () => <mg-popover {...args}>{element}</mg-popover>,
  });

describe('mg-popover', () => {
  test.each([
    { identifier: 'identifier' },
    { identifier: 'identifier', placement: placements[0] },
    { identifier: 'identifier', display: true },
    { identifier: 'identifier', closeButton: true },
    { identifier: 'identifier', display: true },
    { identifier: 'identifier', closeButton: true, lang: 'fr' },
    { identifier: 'identifier', closeButton: true, lang: 'xx' },
  ])('Should render with element', async args => {
    const { root } = await getPage(args, [
      <h2 slot="title">Blu bli blo bla</h2>,
      <p slot="content">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
        exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
        Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </p>,
      <mg-button identifier="identifier-btn">mg-button</mg-button>,
    ]);
    expect(root).toMatchSnapshot();
  });

  test.each([
    { eventIn: 'click', eventOut: 'clickBtn' },
    { eventIn: 'click', eventOut: { code: 'Escape' } },
    { eventIn: 'click', eventOut: 'clickCross' },
    { eventIn: 'click', eventOut: 'clickDocument' },
    { eventIn: 'click', eventOut: 'clickPopover' },
  ])('Should manage display on events %s', async ({ eventIn, eventOut }) => {
    jest.clearAllTimers();

    const args = { identifier: 'identifier', closeButton: true };
    const page = await getPage(args, [
      <h2 slot="title">Blu bli blo bla</h2>,
      <p slot="content">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
        exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
        Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </p>,
      <mg-button identifier="identifier-btn">mg-button</mg-button>,
    ]);

    const mgPopover = page.doc.querySelector('mg-popover');
    const interactiveElement = mgPopover.querySelector(`[aria-controls*='${args.identifier}']`);
    const popover = mgPopover.querySelector(`#${args.identifier}`);
    const popoverButton = popover.querySelector(`mg-button`);

    const displayChangeSpy = jest.spyOn(page.rootInstance.displayChange, 'emit');

    interactiveElement.dispatchEvent(new CustomEvent(eventIn, { bubbles: true }));
    await page.waitForChanges();
    jest.runAllTimers();

    expect(popover).toHaveAttribute('data-show');

    if (typeof eventOut === 'string') {
      if (eventOut === 'clickBtn') {
        interactiveElement.dispatchEvent(new Event('click', { bubbles: true }));
      } else if (eventOut === 'clickCross') {
        popoverButton.dispatchEvent(new Event('click', { bubbles: true }));
      } else if (eventOut === 'clickDocument') {
        document.dispatchEvent(new Event('click', { bubbles: true }));
      } else if (eventOut === 'clickPopover') {
        popover.dispatchEvent(new Event('click', { bubbles: true }));
      }
    } else {
      mgPopover.dispatchEvent(new KeyboardEvent('keydown', { code: eventOut.code }));
    }
    await page.waitForChanges();

    if (eventOut === 'clickPopover') {
      expect(popover).toHaveAttribute('data-show');
      expect(displayChangeSpy).toHaveBeenCalledWith(true);
    } else {
      expect(popover).not.toHaveAttribute('data-show');
      expect(displayChangeSpy).toHaveBeenCalledWith(false);
    }
  });

  test.each([false, true])('Should not toggle display when disabled', async display => {
    const args = { identifier: 'identifier', closeButton: true, disabled: true, display };
    const page = await getPage(args, [
      <h2 slot="title">Blu bli blo bla</h2>,
      <p slot="content">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
        exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.
        Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
      </p>,
      <mg-button identifier="identifier-btn">mg-button</mg-button>,
    ]);

    const mgPopover = page.doc.querySelector('mg-popover');
    const interactiveElement = mgPopover.querySelector(`[aria-controls*='${args.identifier}']`);
    const popover = mgPopover.querySelector(`#${args.identifier}`);
    const popoverButton = popover.querySelector(`mg-button`);

    const displayChangeSpy = jest.spyOn(page.rootInstance.displayChange, 'emit');

    expect(popoverButton).toBeNull();

    if (display) expect(popover).toHaveAttribute('data-show');
    else expect(popover).not.toHaveAttribute('data-show');

    interactiveElement.dispatchEvent(new CustomEvent('click', { bubbles: true }));
    await page.waitForChanges();

    if (display) expect(popover).toHaveAttribute('data-show');
    else expect(popover).not.toHaveAttribute('data-show');

    mgPopover.dispatchEvent(new KeyboardEvent('keydown', { code: 'Escape' }));
    await page.waitForChanges();

    if (display) expect(popover).toHaveAttribute('data-show');
    else expect(popover).not.toHaveAttribute('data-show');

    expect(displayChangeSpy).not.toHaveBeenCalled();
  });

  test('Should throw error if slot title element is not a heading', async () => {
    try {
      const args = { identifier: 'identifier', closeButton: true };
      await getPage(args, [
        <span slot="title">Blu bli blo bla</span>,
        <p slot="content">
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
          exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla
          pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
        </p>,
        <mg-button identifier="identifier-btn">mg-button</mg-button>,
      ]);
    } catch (err) {
      expect(err.message).toContain('<mg-popover> Slotted title must be a heading.');
    }
  });
});
