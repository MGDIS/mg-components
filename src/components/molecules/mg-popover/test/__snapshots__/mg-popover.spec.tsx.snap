// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-popover Should render with element 1`] = `
<mg-popover>
  <mock:shadow-root>
    <slot></slot>
    <div class="mg-popover" id="identifier" style="position: fixed; left: 0; top: 0; margin: 0;">
      <div class="mg-popover__title">
        <slot name="title"></slot>
      </div>
      <slot name="content"></slot>
      <div class="mg-popover__arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
  </mock:shadow-root>
  <h2 slot="title">
    Blu bli blo bla
  </h2>
  <p slot="content">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
  <mg-button aria-controls="identifier" aria-expanded="false" role="button" tabindex="0">
    <mock:shadow-root>
      <div class="mg-button mg-button--primary" id="identifier-btn">
        <div class="mg-button__content">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    mg-button
  </mg-button>
</mg-popover>
`;

exports[`mg-popover Should render with element 2`] = `
<mg-popover>
  <mock:shadow-root>
    <slot></slot>
    <div class="mg-popover" id="identifier" style="position: fixed; left: 0; top: 0; margin: 0;">
      <div class="mg-popover__title">
        <slot name="title"></slot>
      </div>
      <slot name="content"></slot>
      <div class="mg-popover__arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
  </mock:shadow-root>
  <h2 slot="title">
    Blu bli blo bla
  </h2>
  <p slot="content">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
  <mg-button aria-controls="identifier" aria-expanded="false" role="button" tabindex="0">
    <mock:shadow-root>
      <div class="mg-button mg-button--primary" id="identifier-btn">
        <div class="mg-button__content">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    mg-button
  </mg-button>
</mg-popover>
`;

exports[`mg-popover Should render with element 3`] = `
<mg-popover>
  <mock:shadow-root>
    <slot></slot>
    <div class="mg-popover" data-show id="identifier" style="position: fixed; left: 0; top: 0; margin: 0;">
      <div class="mg-popover__title">
        <slot name="title"></slot>
      </div>
      <slot name="content"></slot>
      <div class="mg-popover__arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
  </mock:shadow-root>
  <h2 slot="title">
    Blu bli blo bla
  </h2>
  <p slot="content">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
  <mg-button aria-controls="identifier" aria-expanded="true" role="button" tabindex="0">
    <mock:shadow-root>
      <div class="mg-button mg-button--primary" id="identifier-btn">
        <div class="mg-button__content">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    mg-button
  </mg-button>
</mg-popover>
`;

exports[`mg-popover Should render with element 4`] = `
<mg-popover>
  <mock:shadow-root>
    <slot></slot>
    <div class="mg-popover" id="identifier" style="position: fixed; left: 0; top: 0; margin: 0;">
      <mg-button aria-label="Close" is-icon="" role="button" tabindex="0">
        <mock:shadow-root>
          <div class="mg-button mg-button--flat mg-button--icon">
            <div class="mg-button__content">
              <slot></slot>
            </div>
          </div>
        </mock:shadow-root>
        <mg-icon icon="cross"></mg-icon>
      </mg-button>
      <div class="mg-popover__title">
        <slot name="title"></slot>
      </div>
      <slot name="content"></slot>
      <div class="mg-popover__arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
  </mock:shadow-root>
  <h2 slot="title">
    Blu bli blo bla
  </h2>
  <p slot="content">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
  <mg-button aria-controls="identifier" aria-expanded="false" role="button" tabindex="0">
    <mock:shadow-root>
      <div class="mg-button mg-button--primary" id="identifier-btn">
        <div class="mg-button__content">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    mg-button
  </mg-button>
</mg-popover>
`;

exports[`mg-popover Should render with element 5`] = `
<mg-popover>
  <mock:shadow-root>
    <slot></slot>
    <div class="mg-popover" data-show id="identifier" style="position: fixed; left: 0; top: 0; margin: 0;">
      <div class="mg-popover__title">
        <slot name="title"></slot>
      </div>
      <slot name="content"></slot>
      <div class="mg-popover__arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
  </mock:shadow-root>
  <h2 slot="title">
    Blu bli blo bla
  </h2>
  <p slot="content">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
  <mg-button aria-controls="identifier" aria-expanded="true" role="button" tabindex="0">
    <mock:shadow-root>
      <div class="mg-button mg-button--primary" id="identifier-btn">
        <div class="mg-button__content">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    mg-button
  </mg-button>
</mg-popover>
`;

exports[`mg-popover Should render with element 6`] = `
<mg-popover lang="fr">
  <mock:shadow-root>
    <slot></slot>
    <div class="mg-popover" id="identifier" style="position: fixed; left: 0; top: 0; margin: 0;">
      <mg-button aria-label="Fermer" is-icon="" role="button" tabindex="0">
        <mock:shadow-root>
          <div class="mg-button mg-button--flat mg-button--icon">
            <div class="mg-button__content">
              <slot></slot>
            </div>
          </div>
        </mock:shadow-root>
        <mg-icon icon="cross"></mg-icon>
      </mg-button>
      <div class="mg-popover__title">
        <slot name="title"></slot>
      </div>
      <slot name="content"></slot>
      <div class="mg-popover__arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
  </mock:shadow-root>
  <h2 slot="title">
    Blu bli blo bla
  </h2>
  <p slot="content">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
  <mg-button aria-controls="identifier" aria-expanded="false" role="button" tabindex="0">
    <mock:shadow-root>
      <div class="mg-button mg-button--primary" id="identifier-btn">
        <div class="mg-button__content">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    mg-button
  </mg-button>
</mg-popover>
`;

exports[`mg-popover Should render with element 7`] = `
<mg-popover lang="xx">
  <mock:shadow-root>
    <slot></slot>
    <div class="mg-popover" id="identifier" style="position: fixed; left: 0; top: 0; margin: 0;">
      <mg-button aria-label="Close" is-icon="" role="button" tabindex="0">
        <mock:shadow-root>
          <div class="mg-button mg-button--flat mg-button--icon">
            <div class="mg-button__content">
              <slot></slot>
            </div>
          </div>
        </mock:shadow-root>
        <mg-icon icon="cross"></mg-icon>
      </mg-button>
      <div class="mg-popover__title">
        <slot name="title"></slot>
      </div>
      <slot name="content"></slot>
      <div class="mg-popover__arrow" data-popper-arrow style="position: absolute;"></div>
    </div>
  </mock:shadow-root>
  <h2 slot="title">
    Blu bli blo bla
  </h2>
  <p slot="content">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
  <mg-button aria-controls="identifier" aria-expanded="false" role="button" tabindex="0">
    <mock:shadow-root>
      <div class="mg-button mg-button--primary" id="identifier-btn">
        <div class="mg-button__content">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    mg-button
  </mg-button>
</mg-popover>
`;

exports[`mg-popover Should render with element 8`] = `
<mg-popover>
  <mock:shadow-root>
    <slot></slot>
    <div class="mg-popover" id="identifier" style="position: fixed; left: 0; top: 0; margin: 0;">
      <div class="mg-popover__title">
        <slot name="title"></slot>
      </div>
      <slot name="content"></slot>
      <div data-popper-arrow style="position: absolute;"></div>
    </div>
  </mock:shadow-root>
  <h2 slot="title">
    Blu bli blo bla
  </h2>
  <p slot="content">
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
  </p>
  <mg-button aria-controls="identifier" aria-expanded="false" role="button" tabindex="0">
    <mock:shadow-root>
      <div class="mg-button mg-button--primary" id="identifier-btn">
        <div class="mg-button__content">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    mg-button
  </mg-button>
</mg-popover>
`;
